### PR TITLE
feat: generate breached passwords script

### PR DESCRIPTION
**Pull request by ni-jessica**

### CODE REVIEW

File: backend/database/generatePasswords.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    # generate other characters
+    random_source = string.ascii_letters + string.digits
+    r = random.randint(6,20)
+    for _ in range(r):
+        password += random.choice(random_source)
+
+    # shuffle all characters
+    password_list = list(password)
+    random.SystemRandom().shuffle(password_list)
```
Comment (stellaljung) [inline backend/database/generatePasswords.py:30]: Nice way of mixing up the characters!

File: backend/database/generatePasswords.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    password = ''.join(password_list)
+
+    if password not in passwords:
+        passwords.append(password)
```
Comment (stellaljung) [inline backend/database/generatePasswords.py:34]: If we use a hashset then we could avoid needing to append it to the back since if the value is already in a hash set it will not be added again. But I am not sure if a list is more efficient than a hashset in the user case of having to cryptograph it later.

File: backend/database/generatePasswords.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        i += 1
+
+# write to csv file
+with open('./backend/database/breachedPasswords.csv', 'w', newline='') as csvfile:
+    writer = csv.writer(csvfile, delimiter = '\n')
+    writer.writerow(passwords)
```
Comment (stellaljung) [inline backend/database/generatePasswords.py:40]: Is this why we use a list to append more easily for writing?

File: backend/database/generatePasswords.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import random
+import string
+import csv
+
+'''
+Generates 20 passwords between 10-24 characters that contain:
```
Comment (csirianni) [inline backend/database/generatePasswords.py:6]: **question**: Why 24?

File: backend/database/generatePasswords.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+- at least 1 lowercase letter
+- at least 1 uppercase letter
+- at least 1 digit
+- 1 special character
+
+since when presented with these requirements, most users choose 1 special character and fill in the rest of the characters with letters and digits. Writes to passwords to a csv file.
+'''
+passwords, i = set(), 0
+
+while i < 20:
```
Comment (csirianni) [inline backend/database/generatePasswords.py:16]: **suggestion**: Make this a function: `generate_passwords(num_passwords, min_chars, max_chars) -> set[str]`. Then you don't have to hardcode changes to the values.

File: backend/database/generatePasswords.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    # generate other characters
+    random_source = string.ascii_letters + string.digits
+    r = random.randint(6,20)
+    for _ in range(r):
+        password += random.choice(random_source)
+
+    # shuffle all characters
+    password_list = list(password)
+    random.SystemRandom().shuffle(password_list)
+    password = ''.join(password_list)
```
[the remark this answers is not in the document]
Comment (csirianni) [reply backend/database/generatePasswords.py:32]: `''` is the empty string, so the operation `''.join(password_list)` just combines all the elements in the list into a single word.

File: backend/database/generatePasswords.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+    if password not in passwords:
+        passwords.add(password)
+        i += 1
+
+# write to csv file
+with open('./backend/database/breachedPasswords.csv', 'w', newline='') as csvfile:
+    writer = csv.writer(csvfile, delimiter = '\n')
+    writer.writerow(passwords)
```
Comment (csirianni) [inline backend/database/generatePasswords.py:41]: **suggestion**: Make this a function to with you input being the file path and the set of passwords.

Comment (csirianni) [inline backend/database/generatePasswords.py:41]: **suggestion**: Combine these two functions into a `if __name__ == '__main__'` expression.

Comment (csirianni) [inline backend/database/generatePasswords.py:41]: **suggestion**: Unit test the `generate_numbers()` function in a separate file using `PyTest`.

**Review by csirianni — CHANGES_REQUESTED** [review]
We can make this more exhaustive of user passwords in the future if we want. Maybe by including a string or enum for the type of passwords to generate. But I don't think it impacts the functionality of the system, so it's fine either way.

File: backend/database/generatePasswords.py (hunk 2 above)
Comment (csirianni) [reply backend/database/generatePasswords.py]: Good point.

File: backend/src/main.cpp [hunk 8]
```diff
@@ -5,9 +5,10 @@
 #include "crow.h"
 #include <sqlite3.h>
```
Comment (csirianni) [inline backend/src/main.cpp:6]: **suggestion**: Remove ununused import.

File: backend/src/passwords.cpp [hunk 9]
```diff
@@ -0,0 +1,72 @@
+#include <iostream>
+#include "passwords.hpp"
+
+namespace passwords
```
Comment (csirianni) [inline backend/src/passwords.cpp:4]: **suggestion**: Change namespace (and file) name to `password`, just like `database`.

File: backend/src/passwords.cpp [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+{
+    std::unordered_set<std::string> generatePasswords(const int &numPasswords, const int &maxChars)
```
Comment (csirianni) [inline backend/src/passwords.cpp:6]: **issue**: You don't need to use const reference with primitive data types. Also, parameter names are snake_case, not camelCase.

File: backend/src/passwords.cpp [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    {
+        int minChars = 10;
```
Comment (csirianni) [inline backend/src/passwords.cpp:8]: **issue**: Use snake_case and make this `const``.

File: backend/src/passwords.cpp [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+        if (maxChars < minChars) {
+            std::cout << "Password must have a minimum length of 10 characters\n";
+            exit(0);
+        }
```
Comment (csirianni) [inline backend/src/passwords.cpp:13]: **issue**: This is the C-style way of handling exceptions. C++ adds `throw` and `try` `catch` blocks. Please use that instead.

File: backend/src/passwords.cpp [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+
+        std::unordered_set<std::string> passwords;
+        std::string lowercase = "abcdefghijklmnopqrstuvwxyz";
```
Comment (csirianni) [inline backend/src/passwords.cpp:16]: **suggestion**: Make all these `const` because they don't change.

File: backend/src/passwords.cpp [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+        std::string uppercase = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+        std::string digits = "0123456789";
+        std::string symbols = "!#$%&'()\\*+,-./:;<=>?@[]^_`{|}~\"";
+
+        while (passwords.size() < numPasswords)
+        {
+            // randomize length between minChars and maxChars
+            int length = minChars + rand() % ((maxChars + 1) - minChars);
+            std::string password = ""; 
+            bool hasLowercase = false, hasUppercase = false, hasDigit = false, hasSymbol = false;
```
Comment (csirianni) [inline backend/src/passwords.cpp:26]: **issue**: Don't initialize multiple values in the same line; it makes code harder to maintain/refactor.

File: backend/src/passwords.cpp [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+        std::unordered_set<std::string> passwords;
+        std::string lowercase = "abcdefghijklmnopqrstuvwxyz";
+        std::string uppercase = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+        std::string digits = "0123456789";
+        std::string symbols = "!#$%&'()\\*+,-./:;<=>?@[]^_`{|}~\"";
+
+        while (passwords.size() < numPasswords)
+        {
+            // randomize length between minChars and maxChars
+            int length = minChars + rand() % ((maxChars + 1) - minChars);
```
Comment (csirianni) [inline backend/src/passwords.cpp:24]: **suggestion**: Use `size_t` instead of int since this number can never be negative and represents a size. And make it `const`.

File: backend/src/passwords.hpp [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace passwords
+{
+    /**
+     * @brief Returns a set of passwords, where each password contains at least one lowercase 
+     * letter, at least one uppercase letter, at least one digit, and at least special character.
+     * 
+     * @param numPasswords The number of passwords to generate.
+     * @param maxChars The maximum number of characters for a password.
+     * @return std::unordered_set<std::string>
```
Comment (csirianni) [inline backend/src/passwords.hpp:15]: **suggestion**: Comment the return type too.

File: backend/tests/test.cpp [hunk 17]
```diff
@@ -0,0 +1,29 @@
+#define CATCH_CONFIG_MAIN
+#include <regex>
+#include <catch2/catch_all.hpp>
+#include "../src/passwords.cpp"
+
+std::unordered_set<std::string> passwordSet = passwords::generatePasswords(5, 24);
+
+bool hasLettersAndDigit(std::string str)
```
Comment (csirianni) [inline backend/tests/test.cpp:8]: **suggestion**: This string is const reference. Use a more descriptive parameter name, `password`.

File: backend/tests/test.cpp [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+
+bool hasSymbol(std::string str)
+{
+    return str.find_first_not_of("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ01234567890_") != std::string::npos;
+}
+
+TEST_CASE("Test basic password requirements")
+{
+    // function generated 5 passwords
+    REQUIRE(passwordSet.size() == 5);
```
Comment (csirianni) [inline backend/tests/test.cpp:22]: **issue**: You don't want to use `REQUIRE` here because that stops the program from running the rest of the tests. Use `CHECK` instead which is the typical notion of assert in other languages. For more info, https://github.com/catchorg/Catch2/blob/devel/docs/assertions.md.

File: backend/tests/test.cpp [hunk 19]
```diff
@@ -0,0 +1,29 @@
+#define CATCH_CONFIG_MAIN
```
Comment (csirianni) [inline backend/tests/test.cpp:1]: **question**: Why is this here?

File: backend/tests/test.cpp [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+#include <regex>
+#include <catch2/catch_all.hpp>
+#include "../src/passwords.cpp"
```
Comment (csirianni) [inline backend/tests/test.cpp:4]: **question**: Why not import the header file? I think this should be the header.

File: backend/tests/test.cpp [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+
+std::unordered_set<std::string> passwordSet = passwords::generatePasswords(5, 24);
+
+bool hasLettersAndDigit(std::string str)
+{
+    const std::regex pattern("^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d).+$");
+    return regex_match(str, pattern);
+}
+
+bool hasSymbol(std::string str)
```
Comment (csirianni) [inline backend/tests/test.cpp:14]: **suggestion**: const reference, parameter name.

File: backend/tests/test.cpp [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+{
+    return str.find_first_not_of("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ01234567890_") != std::string::npos;
+}
+
+TEST_CASE("Test basic password requirements")
+{
+    // function generated 5 passwords
+    REQUIRE(passwordSet.size() == 5);
+    // a password has more than 10 characters
+    auto password = *(passwordSet.begin());
```
Comment (csirianni) [inline backend/tests/test.cpp:24]: **issue**: Check this for every password in the set, not just one.

File: backend/tests/test.cpp [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+
+bool hasLettersAndDigit(std::string str)
+{
+    const std::regex pattern("^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d).+$");
+    return regex_match(str, pattern);
+}
+
+bool hasSymbol(std::string str)
+{
+    return str.find_first_not_of("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ01234567890_") != std::string::npos;
```
Comment (csirianni) [inline backend/tests/test.cpp:16]: **praise**: This is an interesting way to do this. My first idea would be `any_of`.

File: backend/tests/test.cpp [hunk 24]
```diff
@@ -0,0 +1,29 @@
+#define CATCH_CONFIG_MAIN
+#include <regex>
+#include <catch2/catch_all.hpp>
+#include "../src/passwords.cpp"
+
+std::unordered_set<std::string> passwordSet = passwords::generatePasswords(5, 24);
```
Comment (csirianni) [inline backend/tests/test.cpp:6]: **issue**: Call this inside the test, not outside.

File: backend/tests/test.cpp [hunk 25]
```diff
@@ -0,0 +1,29 @@
...
+{
+    // function generated 5 passwords
+    REQUIRE(passwordSet.size() == 5);
+    // a password has more than 10 characters
+    auto password = *(passwordSet.begin());
+    REQUIRE(password.length() >= 10);
+    // a password meets all char requirements
+    REQUIRE(hasLettersAndDigit(password));
+    REQUIRE(hasSymbol(password));
+}
```
Comment (csirianni) [inline backend/tests/test.cpp:29]: **issue**: Use a more descriptive name for this file. It could just be `test_password.cpp`.

File: backend/tests/test.cpp [hunk 26]
```diff
@@ -0,0 +1,29 @@
...
+    const std::regex pattern("^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d).+$");
+    return regex_match(str, pattern);
+}
+
+bool hasSymbol(std::string str)
+{
+    return str.find_first_not_of("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ01234567890_") != std::string::npos;
+}
+
+TEST_CASE("Test basic password requirements")
```
Comment (csirianni) [inline backend/tests/test.cpp:19]: **suggestion**: This is really testing `generatePasswords`. Could include that in the description.

File: backend/CMakeLists.txt [hunk 27]
```diff
@@ -16,3 +19,20 @@ add_executable(${PROJECT_NAME} ${SOURCES})
...
+
+set (TEST_SOURCE
+    tests/test.cpp
+)
+
+add_executable(${PROJECT_NAME} ${TEST_SOURCE})
+
+target_link_libraries(${PROJECT_NAME} PRIVATE Catch2::Catch2WithMain)
+
+target_include_directories(${PROJECT_NAME} PRIVATE tests)
```
Comment (csirianni) [inline backend/CMakeLists.txt:36]: **suggestion**: This line isn't necessary because there aren't any header files in `/tests`. If you want to include `passwords.hpp`, make this the `/src` folder.

File: backend/src/passwords.cpp [hunk 28]
```diff
@@ -0,0 +1,72 @@
...
+
+            for (int i = 0; i < length; i++)
+            {
+                int charType = rand() % 4;
+                int index;
+                switch (charType)
+                {
+                case 0: // lowercase letter
+                    hasLowercase = true;
+                    if (!(hasLowercase && hasUppercase && hasDigit && hasSymbol)){
```
Comment (csirianni) [inline backend/src/passwords.cpp:36]: **question**: I don't understand this if statement. If the password is missing a lowercase or an uppercase or a digit or a symbol, stop the loop? What if the password doesn't have the minimum # of characters yet?

File: backend/tests/test.cpp [hunk 29]
```diff
@@ -0,0 +1,29 @@
+#define CATCH_CONFIG_MAIN
+#include <regex>
+#include <catch2/catch_all.hpp>
```
Comment (csirianni) [inline backend/tests/test.cpp:3]: **suggestion**: You can use `#include <catch2/catch_test_macros.hpp>` instead of `catch_all` to reduce the size of the import and thus compilation time.

File: backend/README.md [hunk 30]
```diff
@@ -30,3 +30,9 @@ To fix VS Code import errors with Crow, try adding the following line to your `s
 ```
 
 [Source](https://stackoverflow.com/questions/58077908/linking-conan-include-to-vs-code)
+
+## Testing
+After building, you can run tests from `/backend`:
+```bash
+cd build && ./tests
```
Comment (csirianni) [inline backend/README.md:37]: **thought**: I definitely think it's better to have `make check`. I'll get on that.

File: backend/CMakeLists.txt [hunk 31]
```diff
@@ -16,3 +19,20 @@ add_executable(${PROJECT_NAME} ${SOURCES})
 target_link_libraries(${PROJECT_NAME} Crow::Crow SQLite::SQLite3)
 
 target_include_directories(${PROJECT_NAME} PRIVATE src)
+
+## tests
```
Comment (csirianni) [inline backend/CMakeLists.txt:23]: **thought**: There might be a better way to separate these two things. I'll look into it.

File: backend/CMakeLists.txt [hunk 32]
```diff
@@ -16,3 +19,20 @@ add_executable(${PROJECT_NAME} ${SOURCES})
...
+    tests/test.cpp
+)
+
+add_executable(${PROJECT_NAME} ${TEST_SOURCE})
+
+target_link_libraries(${PROJECT_NAME} PRIVATE Catch2::Catch2WithMain)
+
+target_include_directories(${PROJECT_NAME} PRIVATE tests)
+
+include(Catch)
```
Comment (csirianni) [inline backend/CMakeLists.txt:38]: **question**: Why is this here?

**Review by csirianni — COMMENTED** [review]
Nice work on this. Lots of small details to address, but most of it is not correctness/program logic related. I'll look into the CMakeLists.txt stuff soon.

File: backend/src/passwords.cpp (hunk 28 above)
Comment (ni-jessica) [reply backend/src/passwords.cpp:36]: this just goes to the next iteration of the loop if all the reqs aren't satisfied yet, so i should change it to a while loop

File: backend/tests/test.cpp (hunk 19 above)
Comment (ni-jessica) [reply backend/tests/test.cpp:1]: creates the "main" program to run the tests

File: backend/CMakeLists.txt (hunk 32 above)
Comment (stellaljung) [reply backend/CMakeLists.txt:39]: In the documents for Catch they include this line but do not include the direct reason: 

[Catch Github documents](https://github.com/catchorg/Catch2/blob/devel/docs/cmake-integration.md#top)

Comment (stellaljung) [reply backend/CMakeLists.txt:39]: I am going to delete it though since it is not impacting the running of the backend folder.

**Review by stellaljung — DISMISSED** [review]

File: .vscode/settings.json [hunk 33]
```diff
@@ -0,0 +1,61 @@
...
+        "system_error": "cpp",
+        "tuple": "cpp",
+        "typeinfo": "cpp",
+        "unordered_map": "cpp",
+        "unordered_set": "cpp",
+        "variant": "cpp",
+        "vector": "cpp",
+        "algorithm": "cpp"
+    }
+}
```
Comment (ni-jessica) [inline .vscode/settings.json:61]: issue: can we delete this and add `.vscode` to `.gitignore`

File: backend/src/main.cpp [hunk 34]
```diff
@@ -1,15 +1,23 @@
 #include <stdlib.h>
 #include <stdio.h>
 #include <string.h>
+#include <iostream>
```
Comment (ni-jessica) [inline backend/src/main.cpp:4]: nitpick: delete unused import

File: backend/src/password.cpp [hunk 35]
```diff
@@ -0,0 +1,86 @@
...
+        try {
+            if (max_chars < min_chars)
+            {
+                throw (max_chars);
+            }
+        }
+        catch (int max_chars) {
+            std::cout << "Password must have a minimum length of 10 characters\n";
+            std::cout << "You inputted a maximum length of " << max_chars << " characters \n";
+        }
```
Comment (ni-jessica) [inline backend/src/password.cpp:20]: praise: amazing error handling!

File: backend/tests/testPasswordGenerator.cpp [hunk 36]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+bool hasSymbol(const std::string &password)
+{
+    return password.find_first_not_of("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ01234567890") != std::string::npos;
+}
+
+TEST_CASE("Test generatePassword creates valid passwords")
+{
+    std::unordered_set<std::string> passwordSet = password::generatePasswords(3, 12);
```
Comment (ni-jessica) [inline backend/tests/testPasswordGenerator.cpp:19]: nitpick: this was my bad but I forgot to change this to snake case 😅

**Comment by ni-jessica** [comment]
maybe we want to split CMakeList.txt into subdirectories https://stackoverflow.com/questions/42744315/cmake-with-subdirectories#:~:text=One%20CMakeLists.txt%20file%20in%20each%20directory%2C%20each%20one%20brought%20in%20by%20its%20parent%20using%20add_subdirectory().%20This%20approach%20is%20very%20common%20and%20is%20usually%20the%20recommended%20one.

File: backend/src/password.cpp [hunk 37]
```diff
@@ -0,0 +1,86 @@
...
+            bool has_symbol = false;
+
+            while (password.length() < length)
+            {
+                int char_type = rand() % 4;
+                int index;
+                switch (char_type)
+                {
+                case 0: // lowercase letter
+                    if (has_lowercase && !(has_uppercase && has_digit && has_symbol))
```
Comment (csirianni) [inline backend/src/password.cpp:45]: **praise**: I like this exception! I get it now.

File: backend/tests/testPasswordGenerator.cpp [hunk 38]
```diff
@@ -0,0 +1,29 @@
...
+    // function generated 3 passwords
+    CHECK(password_set.size() == 3);
+
+    for (const auto &password : password_set)
+    {   
+        CHECK(password.length() >= 10);
+        CHECK(hasLettersAndDigit(password));
+        CHECK(hasSymbol(password));
+    }
+}
```
Comment (csirianni) [inline backend/tests/testPasswordGenerator.cpp:29]: **praise**: Perfect!

**Review by csirianni — CHANGES_REQUESTED** [review]
Nice job! Thanks for making all those changes. Just a couple more comments.

**Comment by csirianni** [comment]
> maybe we want to split CMakeList.txt into subdirectories https://stackoverflow.com/questions/42744315/cmake-with-subdirectories#:~:text=One%20CMakeLists.txt%20file%20in%20each%20directory%2C%20each%20one%20brought%20in%20by%20its%20parent%20using%20add_subdirectory().%20This%20approach%20is%20very%20common%20and%20is%20usually%20the%20recommended%20one.

Yup, agreed. I will do this after my remaining comments are resolved.

**Review by csirianni — APPROVED** [review]
